libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1027 @@
+{
+  FlatpakRemoteRefAction **action_a = (FlatpakRemoteRefAction **) a;
+  FlatpakRemoteRefAction **action_b = (FlatpakRemoteRefAction **) b;
+
+  return (gint) (*action_a)->serial - (gint) (*action_b)->serial;
+}
+
+static JsonNode *
+parse_json_from_file (GFile         *file,
+                      GCancellable  *cancellable,
+                      GError       **error)
+{
+  g_autoptr(GFileInputStream) input_stream = NULL;
+  g_autoptr(JsonParser) parser = json_parser_new_immutable ();
+
+  input_stream = g_file_read (file, cancellable, error);
+
+  if (!input_stream)
+    return NULL;
+
+  if (!json_parser_load_from_stream (parser,
+                                     G_INPUT_STREAM (input_stream),
+                                     cancellable,
+                                     error))
+    return NULL;
+
+  return json_node_ref (json_parser_get_root (parser));
+}
+
+static const gchar *
+eos_updater_get_system_architecture_string (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_OVERRIDE_ARCHITECTURE",
+                                    flatpak_get_default_arch ());
+}
+
+static gboolean
+architecture_skip_filter_applies (JsonNode  *filter_value,
+                                  gboolean  *is_filtered,
+                                  GError   **error)
+{
+  JsonArray *skip_architectures_array = json_node_get_array (filter_value);
+  g_autoptr(GList) architecture_nodes = NULL;
+  g_autofree gchar *node_str = json_node_to_string (filter_value);
+  GList *iter = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (skip_architectures_array == NULL)
+    {
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected '~architectures' filter to be an array, was: %s",
+                   node_str);
+      return FALSE;
+    }
+
+  architecture_nodes = json_array_get_elements (skip_architectures_array);
+
+  for (iter = architecture_nodes; iter != NULL; iter = iter->next)
+    {
+      const gchar *string = json_node_get_string (iter->data);
+
+      if (string == NULL)
+        {
+          g_set_error (error,
+                       EOS_UPDATER_ERROR,
+                       EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                       "Expected '~architectures' array to contain string values, was: %s",
+                       node_str);
+          return FALSE;
+        }
+
+      if (g_strcmp0 (string, eos_updater_get_system_architecture_string ()) == 0)
+        {
+          *is_filtered = TRUE;
+          return TRUE;
+        }
+    }
+
+  return TRUE;
+}
+
+static gboolean
+action_filter_applies (JsonObject   *object,
+                       const gchar  *filter_key_name,
+                       gboolean     *is_filtered,
+                       GError      **error)
+{
+  g_autoptr (JsonNode) filter_value = json_object_get_member (object, filter_key_name);
```

Comment:
Should not be `g_autoptr()` since `json_object_get_member()` does not return a new ref.